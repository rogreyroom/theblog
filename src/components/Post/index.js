import React, { useState, useEffect } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import PropTypes from 'prop-types'
import { favsActions } from '../../redux/favs/favSlice'
import { SvgFavNo } from '../icons/FavNoIcon'
import { SvgFavYes } from '../icons/FavYesIcon'
import classes from './index.module.scss'

const Post = ({ userName, photoUrl, photoTitle, postTitle, postBody }) => {
  const [favPost, setFavPost] = useState('no')
  const { post } = useSelector((state) => state.posts)
  const { fav } = useSelector((state) => state.favs)
  const dispatch = useDispatch()

  useEffect(() => {
    const isFav = fav.find((favId) => favId === post.id.toString())
    return isFav ? setFavPost(() => 'yes') : setFavPost(() => 'no')
  }, [post, fav])

  const toggleFav = (id) =>
    favPost === 'no'
      ? dispatch(favsActions.addFav(id))
      : dispatch(favsActions.removeFav(id))

  return (
    <>
      <article className={classes.post}>
        <div className={classes.post__info}>
          <span className={classes.post__user}>{userName}</span>
          <button
            type="button"
            onClick={() => toggleFav(post.id)}
            className={[
              classes.post__favBtn,
              classes[`post__favBtn--${favPost}`],
            ].join(' ')}
          >
            {favPost === 'no' ? <SvgFavNo /> : <SvgFavYes />}
          </button>
        </div>
        <img src={photoUrl} alt={photoTitle} className={classes.post__photo} />
        <h2 className={classes.post__title}>{postTitle}</h2>
        <p className={classes.post__text}>{postBody}</p>
      </article>
    </>
  )
}

Post.propTypes = {
  userName: PropTypes.string.isRequired,
  photoUrl: PropTypes.string.isRequired,
  photoTitle: PropTypes.string.isRequired,
  postTitle: PropTypes.string.isRequired,
  postBody: PropTypes.string.isRequired,
}

export default Post
