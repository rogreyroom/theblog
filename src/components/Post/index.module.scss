.post {
	display: grid;
	grid-template-columns: clamp(17.75rem, -1.9821rem + 92.8571vw, 50.25rem);
	grid-template-rows: min-content 1fr max-content auto;
	grid-template-areas: 'info' 'photo' 'title' 'body';
	margin-bottom: clamp(1rem, -0.1034rem + 5.5172vw, 3rem);

	&__info {
		grid-area: info;
		margin: 0;
		display: flex;
		flex-direction: row;
		justify-content: space-between;
	}

	&__user {
		color: var(--c-grey);
		font-size: clamp(var(--fs-text-small), 0.8060rem + 0.3448vw, var(--fs-text-normal));
		margin: 0;
	}

	&__photo {
		grid-area: photo;
		background-color: var(--c-grey-alpha);
		margin: 0;
		max-width: clamp(17.75rem, -1.9821rem + 92.8571vw, 50.25rem);
		max-height: clamp(7.5rem, -2.8448rem + 51.7241vw, 26.25rem);
		width: 100%;
		object-fit: cover;
	}

	&__title {
		grid-area: title;
		color: var(--c-black);
		font-size: clamp(var(--fs-header-small), 0.4483rem + 2.7586vw, var(--fs-header-large));
		display: block;
		background-color: var(--c-grey-alpha);
		margin: 0;
		padding: clamp(var(--m), 0.0905rem + 5.1724vw, var(--xxl));
		text-align: center;
	}

	&__text {
		grid-area: body;
		color: var(--c-black);
		font-size: clamp(var(--fs-text-normal), 0.7212rem + 0.7692vw, var(--fs-text-large));
		margin: var(--normal) 0 0 0;
		padding: clamp(var(--xs), 0.2241rem + 1.3793vw, var(--normal));
	}

	&__favBtn {
		--size: clamp(var(--normal), 0.7241rem + 1.3793vw, var(--l));
    --red: hsla(350, 94%, 48%, 94%);
		grid-area: fav;
		display: flex;
		background: transparent;
		box-shadow: none;
		color: var(--c-black);
		font-family: inherit;
		font-size: var(--size);
		line-height: var(--size);
		text-align: center;
		text-decoration: none;
		border: none;
		margin: 0;
		padding: 0;
		-webkit-appearance: none;
		-moz-appearance: none;
		cursor: pointer;
		transition: all 250ms ease-in-out;
		justify-content: center;
		align-items: center;
		max-width: var(--xxl);
		max-height: var(--xxl);

		& > svg {
			width: var(--size);
			height: var(--size);
			margin: 0;
		}

		&--yes {
			color: var(--red);
			&:hover > svg {
				filter: drop-shadow(0px 0px 6px var(--c-accent));
				color: var(--c-accent);

				&:focus,
				&:active {
					outline: 3px solid transparent;
					box-shadow: 0 0 1px 2px var(--c-red);
					filter: drop-shadow(0px 0px 6px var(--red));
				}
			}
		}

		&--no {
			color: var(--c-accent);
			&:hover > svg {
				filter: drop-shadow(0px 0px 6px var(--red));
				color: var(--red);

				&:focus,
				&:active {
					outline: 3px solid transparent;
					box-shadow: 0 0 1px 2px var(--c-accent);
					filter: drop-shadow(0px 0px 6px var(--red));
				}
			}
		}
	}
}
