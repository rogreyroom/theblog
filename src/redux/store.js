import { configureStore } from '@reduxjs/toolkit'
import postsSlice from './posts/postsSlice'
import commentsSlice from './comments/commentsSlice'

const store = configureStore({
  reducer: {
    posts: postsSlice.reducer,
    comments: commentsSlice.reducer,
  },
})

export default store
