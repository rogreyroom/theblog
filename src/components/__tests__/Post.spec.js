/**
 * @jest-environment jsdom
 */
import React from 'react'
import { screen, render, cleanup } from '@testing-library/react'
import { Provider } from 'react-redux'
import Post from '../Post'
import store from '../../redux/store'

describe('PostCard component', () => {
  let element
  beforeEach(() => {
    const { container } = render(
      <Provider store={store}>
        <Post
          userName="John"
          photoUrl="https://via.placeholder.com/600/92c952"
          photoTitle="placeholder photo"
          postTitle="Awesome blog title"
          postBody="Some cool blog text"
        />
      </Provider>,
    )
    element = container
  })

  it('renders the component', () => {
    expect(element).toMatchSnapshot()
  })

  it('component contains user name "John"', () => {
    expect(screen.getByText(/john/i)).toHaveTextContent('John')
  })

  it('component contains photo url "https://via.placeholder.com/600/92c952"', () => {
    expect(element.querySelector('img').src).toEqual(
      'https://via.placeholder.com/600/92c952',
    )
  })

  it('component contains post title "Awesome blog title"', () => {
    expect(screen.getByText(/awesome/i)).toHaveTextContent('Awesome blog title')
  })

  it('component contains post body "Some cool blog text"', () => {
    expect(screen.getByText(/cool/i)).toHaveTextContent('Some cool blog text')
  })

  it('component contains svg icon', () => {
    expect(element.querySelector('svg')).toBeInTheDocument()
  })

  afterAll(cleanup)
})
