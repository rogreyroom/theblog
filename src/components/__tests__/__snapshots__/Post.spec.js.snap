// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostCard component renders the component 1`] = `
<div>
  <article
    class="post"
  >
    <div
      class="post__info"
    >
      <span
        class="post__user"
      >
        John
      </span>
      <button
        class="post__favBtn post__favBtn--no"
        type="button"
      >
        <svg
          fill="none"
          height="1em"
          viewBox="0 0 32 32"
          width="1em"
        >
          <path
            d="M23.2 0C20.416 0 17.744 1.296 16 3.344C14.256 1.296 11.584 0 8.8 0C3.872 0 0 3.872 0 8.8C0 14.848 5.44 19.776 13.68 27.264L16 29.36L18.32 27.248C26.56 19.776 32 14.848 32 8.8C32 3.872 28.128 0 23.2 0ZM16.16 24.88L16 25.04L15.84 24.88C8.224 17.984 3.2 13.424 3.2 8.8C3.2 5.6 5.6 3.2 8.8 3.2C11.264 3.2 13.664 4.784 14.512 6.976H17.504C18.336 4.784 20.736 3.2 23.2 3.2C26.4 3.2 28.8 5.6 28.8 8.8C28.8 13.424 23.776 17.984 16.16 24.88Z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <img
      alt="placeholder photo"
      class="post__photo"
      src="https://via.placeholder.com/600/92c952"
    />
    <h2
      class="post__title"
    >
      Awesome blog title
    </h2>
    <p
      class="post__text"
    >
      Some cool blog text
    </p>
  </article>
</div>
`;
